### DISCUSSION TITLE
Using RTK Query, Supabase, and NextJS

### DISCUSSION THREAD

**Discussion by michaelckelly** [supabase/supabase#10199] [Questions]
I'm trying to use Redux Toolkit Query, NextJS, and Supabase together. 

**tl;dr** How do you access the authenticated Supabase browser client outside of a React component?

More details below...

Prior to `@supabase/supabase-js@2.x.x` and `@supabase/auth-helpers-nextjs@0.5.x` this worked out-of-the-box with row level security enabled. Once I migrated to `2.x.x` and `0.5.x`, the following broke:

**supabaseClient.js**

```js
import { createClient } from '@supabase/supabase-js'

const supabaseUrl = process.env.NEXT_PUBLIC_SUPABASE_URL
const supabaseAnonKey = process.env.NEXT_PUBLIC_SUPABASE_ANON_KEY

export const supabase = createClient(supabaseUrl, supabaseAnonKey)
```

**supabaseApi.js**

```js
import { createApi, fetchBaseQuery } from '@reduxjs/toolkit/query/react';

// Import Supabase client
// Prior to 0.5.x, RLS "automagically" worked once session was set
import { supabase } from '../lib/supabaseClient'

export const supabaseApi = createApi({
  tagTypes: ['Item'],
  baseQuery: fetchBaseQuery({ baseUrl: '/api' }),
  endpoints: (builder) => ({
    getItems: builder.query({
      queryFn: async (userId) => {
        let { data, error, status } = await supabase
          .from('items')
          .select(`*`)
          .eq('user_id', userId);
        return { data };
      },
    }),
  })
});

export const {
  useGetItemsQuery,
} = supabaseApi;
```

Specifically, the supabase client seemed to "lose" the correct auth data somehow so the queries were silently returning no data because the RLS policies were not met. 

My initial workaround was to pass the Supabase client from the React component to the RTK `queryFn` function, along the lines of:

**React Component**
```js
  const { data: items = [], isLoading: isLoadingItems } = useGetItemsQuery({
    userId: props.user.id,
    options: { supabase: supabaseClient }
  });
```

This works but is unwieldy because I have to pass the supabase client every time I call the query AND Redux doesn't like the Supabase client in the store.

What is the best practice for initializing a Supabase client in the browser (user has authenticated) outside of a React component (so I can't use the `useSupabase` hook)? Is there a way to do this without needing to pass the supabase client with every API call that I make.

Is there a best practice for keeping a single instance of a supabase client around in a React/Redux app? I see it provided as a prop to `SessionContextProvider` and that seemingly enables the hooks to create a client / access the logged in user. Is it ever stored in a redux store? 

**tl;dr** How do you access the authenticated Supabase browser client outside of a React component?

**Comment by skibitsky** [comment]
Hey @michaelckelly, did you find a solution?

**Comment by cstayyab** [comment]
Anyone found any solution? It would be great if someone from Supabase Team or one of the maintainers can provide a detailed solution to this problem.
